### PR TITLE
Add the `try_trait_v2` library basics

### PR DESCRIPTION
**Pull request by scottmcm**
No compiler changes as part of this -- just new unstable traits and impls thereof.

The goal here is to add the things that aren't going to break anything, to keep the feature implementation simpler in the next PR.

(Draft since the FCP won't end until Saturday, but I was feeling optimistic today -- and had forgotten that FCP was 10 days, not 7 days.)

### CODE REVIEW

**Comment by scottmcm** [comment]
Updated with a real tracking issue, now that the RFC is merged, so I think this is good to go now.

(Sorry for not **ghost**ing it previously.)

File: library/core/src/ops/control_flow.rs [hunk 1]
```diff
@@ -81,6 +82,35 @@ impl<B, C> Try for ControlFlow<B, C> {
     }
 }
 
+#[unstable(feature = "try_trait_v2", issue = "84277")]
+impl<B, C> ops::TryV2 for ControlFlow<B, C> {
+    type Output = C;
+    type Residual = ControlFlow<B, convert::Infallible>;
```
Comment (yaahc) [inline library/core/src/ops/control_flow.rs:88]: What's the reasoning for using `Infallible` over `!`?

File: library/core/src/ops/mod.rs [hunk 2]
```diff
@@ -184,6 +185,12 @@ pub use self::range::{Bound, RangeBounds, RangeInclusive, RangeToInclusive};
 #[unstable(feature = "try_trait", issue = "42327")]
 pub use self::r#try::Try;
 
+#[unstable(feature = "try_trait_v2", issue = "84277")]
+pub use self::try_trait::FromResidual;
+
+#[unstable(feature = "try_trait_transition", reason = "for bootstrap", issue = "none")]
```
Comment (yaahc) [inline library/core/src/ops/mod.rs:191]: :+1:

File: library/core/src/ops/try_trait.rs [hunk 3]
```diff
@@ -0,0 +1,225 @@
+use crate::ops::ControlFlow;
+
+/// The trait used for a variety of operations related to short-circuits,
+/// such as the `?` operator, `try {}` blocks, and `try_*` methods.
```
Comment (yaahc) [inline library/core/src/ops/try_trait.rs:4]: Not in love with this description of the Try trait but having trouble thinking of anything better. I normally describe the try trait as "modeling control flow, specifically early returns", but I'm not confident this will get the point across more effectively than your current iteration. Interested in your thoughts.

File: library/core/src/ops/try_trait.rs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+///
+/// # Using `Try` in Generic Code
```
Comment (yaahc) [inline library/core/src/ops/try_trait.rs:6]: Might be worth adding another example detailing how to implement a custom `Try` type, to give an idea of the motivating usecases.

File: library/core/src/ops/try_trait.rs [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+///     mut accum: A,
+///     mut f: impl FnMut(A, T) -> R,
+/// ) -> R {
+///     todo!()
+/// }
+/// ```
+///
+/// `Try` is also the trait we need to get the updated accumulator from `f`'s return
+/// value and return the result if we manage to get through the entire iterator:
+/// ```
```
Comment (yaahc) [inline library/core/src/ops/try_trait.rs:47]: Had trouble parsing this step of the example

File: library/core/src/ops/try_trait.rs [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+///             ControlFlow::Break(r) => return R::from_residual(r),
+///         }
+///     }
+///     R::from_output(accum)
+/// }
+/// ```
+///
+/// But this "call `branch`, then `match` on it, and `return` if it was a
+/// `Break`" is exactly what happens inside the `?` operator.  So rather than
+/// do all this manually, we can just use `?` instead:
```
Comment (yaahc) [inline library/core/src/ops/try_trait.rs:95]: Oooo this is soo satisfying.

File: library/core/src/ops/try_trait.rs [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    /// The choice of this type is critical to interconversion.
+    /// Unlike the `Output` type, which will often be a raw generic type,
+    /// this type is typically a newtype of some sort to "color" the type
+    /// so that it's distinguishable from the residuals of other types.
+    ///
+    /// This is why `Result<T, E>::Residual` is not `E`, but `Result<Infallible, E>`.
+    /// That way it's distinct from `ControlFlow<E>::Residual`, for example,
+    /// and thus `?` on `ControlFlow` cannot be used in a method returning `Result`.
+    ///
+    /// In a type that's generic on a parameter that's used as the `Output` type,
```
Comment (yaahc) [inline library/core/src/ops/try_trait.rs:134]: Hard to parse

File: library/core/src/ops/try_trait.rs [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    /// call it `Foo<T> : Try` where `Foo<T>::Output == T`, it's typically easiest
+    /// to make the corresponding `Residual` type by filling in that generic
+    /// with an uninhabited type: `type Residual = Foo<Infallible>;`.
+    #[unstable(feature = "try_trait_v2", issue = "84277")]
+    type Residual;
+
+    /// Wraps up a value such that `?` on the value will produce the original value.
```
Comment (yaahc) [inline library/core/src/ops/try_trait.rs:141]: It seems strange to me to describe this method in terms of one of its properties. I would prefer something like "Constructs a `Try` type from it's `Output` type."

File: library/core/src/ops/try_trait.rs [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+/// Used to specify which residuals can be converted into which [`Try`] types.
+///
+/// Every `Try` type needs to be recreatable from its own associated
+/// `Residual` type, but can also have additional `FromResidual` implementations
+/// to support interconversion with other `Try` types.
+#[unstable(feature = "try_trait_v2", issue = "84277")]
+pub trait FromResidual<R = <Self as Try>::Residual> {
+    /// Produces the return value of the function from the residual
+    /// when the `?` operator results in an early exit.
```
Comment (yaahc) [inline library/core/src/ops/try_trait.rs:207]: Similar to my comment on `from_output`, I would prefer to describe this as a constructor, though I'm less sure on this one because of how it interacts with the Residual wrapper.

**Review by yaahc — COMMENTED** [review]
Very exciting!

File: library/core/src/ops/control_flow.rs (hunk 1 above)
Comment (scottmcm) [reply library/core/src/ops/control_flow.rs:88]: I don't want to accidentally introduce (another) way to refer to real-`!` if `never_type` hasn't stabilized yet but someone gets a hold of this residual type's name.  Like if someone finds a way to use a trick like
```
trait Foo { type Bar; }
impl<T, E> Foo for Result<T, E> { type Bar = T; }
type RealNever = <Result<(),()>::Residual as Foo>::Bar;
```
(I don't know how they'd mention `::Residual` like that, but with inference I worry they'd find a clever way.)

Hopefully the work on never type stabilization will mean soon that this just becomes `!` before that can become an issue, but I've added a checkbox in https://github.com/rust-lang/rust/issues/84277#issue-860487457 to make sure we look at this before stabilizing `try_trait_v2`.

**Comment by Johannesd3** [comment]
I'm not sure whether this is the right place to ask, but were there ever any thoughts to add the following impl?

```rust
impl<B> FromResidual<ControlFlow<B, convert::Infallible>> for B { ... }
```

to make something like this possible:

```rust
fn do_something() -> ControlFlow<()> { ... }

fn main() {
    do_something()?;
    do_something()?;
    do_something()?;
}
```
or would it be a stupid idea and bad use of the `?` operator?

**Comment by scottmcm** [comment]
@Johannesd3 This PR is just adding things from the [approved RFC](https://rust-lang.github.io/rfcs/3058-try-trait-v2.html#reference-level-explanation), so this isn't the place to discuss additional `impl`s.

I'd point you to [IRLO](https://internals.rust-lang.org/) if you'd like to start a conversation about the desirability of that one.  Some things you might want to consider: how such a blanket impl would interact with the coherence rules, and whether potentially having `try{}` blocks in the future might impact how code like this would be written.

File: library/core/src/ops/try_trait.rs (hunk 3 above)
Comment (scottmcm) [reply library/core/src/ops/try_trait.rs:4]: Yeah, I'm not sure either.  Based on <https://doc.rust-lang.org/nightly/std/ops/index.html#traits> I've updated the first line to just talk about the "operators" that use this, and have added some more text before the example.

File: library/core/src/ops/try_trait.rs (hunk 4 above)
Comment (scottmcm) [reply library/core/src/ops/try_trait.rs:6]: More docs are definitely good, but I think I'll leave that for a future PR, if that's ok.

File: library/core/src/ops/try_trait.rs (hunk 7 above)
Comment (scottmcm) [reply library/core/src/ops/try_trait.rs:134]: Rewrote this bit to use more code and fewer words; hopefully it's better.

File: library/core/src/ops/try_trait.rs (hunk 8 above)
Comment (scottmcm) [reply library/core/src/ops/try_trait.rs:141]: Updated both of these to the "constructor" terminology, and moved the expected semantics to a different paragraph.

File: library/core/src/ops/try_trait.rs (hunk 7 above)
Comment (yaahc) [reply library/core/src/ops/try_trait.rs]: definitely much easier to follow, thank you

File: library/core/src/ops/try_trait.rs (hunk 3 above)
Comment (yaahc) [reply library/core/src/ops/try_trait.rs]: that seems reasonable, I never noticed the ops traits were all described that way.

**Comment by yaahc** [comment]
This PR looks good to me. I would like to see the bullet on further planned documentation updates recorded somewhere, presumably on the tracking issue.

**Comment by scottmcm** [comment]
It had the standard "adjust documentation" checkbox, but that's often more about the reference and such, so I've added another one for more library documentation in https://github.com/rust-lang/rust/issues/84277.

**Comment by yaahc** [comment]
Since the RFC has merged am I right in assuming this is ready to merge now?

**Comment by yaahc** [comment]
yolo

@rfcbot merge

**Comment by scottmcm** [comment]
As this is all unstable I don't think it needs an FCP?  It should be entirely invisible to stable code.

**Comment by m-ou-se** [comment]
Looks great to me!

Also verified that this only adds unstable things and has no effect on existing code.

@bors r=yaahc,m-ou-se
